go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,336 @@
+	"errors"
+	"fmt"
+	"github.com/Fantom-foundation/Carmen/go/common"
+	"github.com/Fantom-foundation/Carmen/go/common/amount"
+	"github.com/Fantom-foundation/Carmen/go/common/interrupt"
+	"github.com/Fantom-foundation/Carmen/go/common/witness"
+	"golang.org/x/exp/maps"
+	"math/rand"
+)
+
+// VerifyProofArchiveTrie verifies the consistency of witness proofs for an archive trie.
+// It reads the trie for each block within the input range.
+// It gets account and storage slots and extracts a witness proof for each account and its storage.
+// It is checked that values in the database and the proof match.
+func VerifyProofArchiveTrie(ctx context.Context, dir string, config MptConfig, from, to int, observer VerificationObserver) error {
+	trie, err := OpenArchiveTrie(dir, config, NodeCacheConfig{})
+	if err != nil {
+		return err
+	}
+
+	observer.StartVerification()
+	err = errors.Join(
+		verifyProofArchiveTrie(ctx, trie, from, to, observer),
+		trie.Close())
+	observer.EndVerification(err)
+	return err
+}
+
+func verifyProofArchiveTrie(ctx context.Context, trie *ArchiveTrie, from, to int, observer VerificationObserver) error {
+	blockHeight, empty, err := trie.GetBlockHeight()
+	if err != nil {
+		return err
+	}
+	if empty {
+		return nil
+	}
+
+	if from < 0 || from > int(blockHeight) {
+		from = 0
+		observer.Progress(fmt.Sprintf("adjusting to a minimum block height: %d", from))
+	}
+
+	if to <= 0 || to > int(blockHeight) {
+		to = int(blockHeight) + 1
+		observer.Progress(fmt.Sprintf("adjusting to a maximum block height: %d", to))
+	}
+
+	observer.Progress(fmt.Sprintf("Verifying total %d blocks", to-from))
+	for i := uint64(from); i < uint64(to); i++ {
+		trie, err := trie.getView(i)
+		if err != nil {
+			return err
+		}
+
+		observer.Progress(fmt.Sprintf("Verifying blcok: %d ", i))
+		if err := verifyProofLiveTrie(ctx, trie, observer); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// VerifyProofLiveTrie verifies the consistency of witness proofs for a live trie.
+// It reads the trie for the head block and loads accounts and storage slots.
+// It extracts witness proofs for these accounts and its storage,
+// and checks that values in the proof and the database match.
+func VerifyProofLiveTrie(ctx context.Context, dir string, config MptConfig, observer VerificationObserver) error {
+	trie, err := OpenFileLiveTrie(dir, config, NodeCacheConfig{})
+	if err != nil {
+		return err
+	}
+
+	observer.StartVerification()
+	err = errors.Join(
+		verifyProofLiveTrie(ctx, trie, observer),
+		trie.Close())
+	observer.EndVerification(err)
+	return err
+}
+
+func verifyProofLiveTrie(ctx context.Context, trie *LiveTrie, observer VerificationObserver) error {
+	rootHash, hints, err := trie.UpdateHashes()
+	if hints != nil {
+		hints.Release()
+	}
+	if err != nil {
+		return err
+	}
+
+	observer.Progress("Collecting states... ")
+	// collect all accounts for this block
+	visitor := accountCollectingVisitor{accounts: accountsMap{}, ctx: ctx}
```

Comment:
Running the collector here requires a lot of memory, since it is essentially loading the entire state DB into memory.

To be scalable for large DBs, the visiting of accounts should be merged with the checking of proofs. You run the visitior, and for every account you encounter, you run the proof extraction inside the visitor and check the results. Then you load the slots and check those as well (this should be Ok, although there are accounts with >1GiB of storage); Once the account-iterator is moving on, all state is abandoned.

Reply:
changed this way, thanks for this feedback